### PR TITLE
Apply strict HMAC and KMAC keysize check in FIPS mode.

### PR DESCRIPTION
**Pull request by slontis**
A separate 'internal' query table has been added to the FIPS provider
for MACS. Giving HMAC and KMAC a separate dispatch table allows KDF's to ignore
the key check. If a KDF requires the key check then it must do the
check itself. The normal MAC dispatch table would be used if a user just
fetches HMAC directly.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Related to https://github.com/openssl/openssl/pull/24199

**Comment by slontis** [comment]
@xnox Hopefully https://github.com/openssl/openssl/pull/23900 will get merged tomorrow.
Do you want to continue your PR, or can I apply your code to this PR in some form..

**Comment by slontis** [comment]
We already have keys for OSSL_MAC_PARAM_KEY and OSSL_MAC_PARAM_SALT but unfortunately we cant use them both without breaking compatability with existing code/providers (Since the key is present directly in both EVP_MAC_init and EVP_Q_mac)

**Comment by paulidale** [comment]
> We already have keys for OSSL_MAC_PARAM_KEY and OSSL_MAC_PARAM_SALT but unfortunately we cant use them both without breaking compatability with existing code/providers (Since the key is present directly in both EVP_MAC_init and EVP_Q_mac)

+100 for this resolution.

**Comment by slontis** [comment]
This still has a problem if a user uses the bypass then it wont have an indicator - and there is not much I can do about it.

**Comment by xnox** [comment]
hmac with key-is-salt is unapproved and should raise indicator when said value is returned to the user.
HKDF which uses the above is approved, thus when the above interim value is returned - can KDF clear the indicator?

Didn't we have similar issue with other types of KDF - as in TLS md5/sha1 KDF usage was approved before, but direct usage of md5 was not? Or is it exactly the same one?

**Comment by slontis** [comment]
The trick here is that the indicator for HMAC would need to take into account the flag.
The callback would not get triggered however if the flag is set.
Higher level algorithms have their own indicator, and need to make the decision about what their indicator tests.

**Comment by slontis** [comment]
@paulidale I wonder if there could be some way to have an internal fetchable algorithm that is not visible in core, and then have a HMAC-WITH-SALT algorithm?

**Comment by xnox** [comment]
I would appreciate separate algorithm too such as HMAC-WITH-SALT. And then it is very easy to reason about it in security policies too. "hmac allowed", "hmac-with-salt unallowed internal only exposed API" (see for example boringcrypto security policy with many internal apis listed as exposed and unapproved).

**Comment by paulidale** [comment]
> Didn't we have similar issue with other types of KDF - as in TLS md5/sha1 KDF usage was approved before, but direct usage of md5 was not?

This is different.  MD5 is included inside the FIPS provider but it isn't exposed to users of the module.  I.e. it's not listed in the table of digests that libcrypto gets given.  It's impossible to use the FIPS provider's MD5 code for anything except the KDF (where it's allowed).

**Comment by paulidale** [comment]
> @paulidale I wonder if there could be some way to have an internal fetchable algorithm that is not visible in core, and then have a HMAC-WITH-SALT algorithm? Or an internal settable?

@slontis and I had a long discussion about this and came up with a number of possible solutions:
1. Have a different table of MAC algorithms internally compared to the exported list.  Use this table for the internal library context.
2. Implement an internal provider that is loaded into the FIPS provider's internal library context.  This provider containing alternate implementations of algorithms that don't include the indicator support.  They'd not be externally visible and could only be called as part of higher level algorithms and such use is automatically approved.
3. Implement an alternative MAC algorithm and build an (internal) EVP_MAC structure for it which is then used directly by the KDF or whatever.  Again, not use visible.
4. Add the concept of algorithm _usage_ via a settable OSSL_PARAM.  This would be user facing (& thus open to abuse) but it would address a number of other concerns.  E.g. MAC being used for verification has different FIPS restrictions than one being used for signing.
5. Add an internal function which sets an indicator bypass flag on the MAC's context.  This is a bit ugly since it breaks the compartmentalisation of the various objects, but it's likely the easiest to implement.

1 & 2 are conceptually quite similar.  We could even go as far as not loading the FIPS provider's _external_ function list into it's internal library context.  We know precisely which algorithms can be used by approved higher level ones and only these would need to be included for internal use.

**Comment by slontis** [comment]
I have implemented the internal table in this PR. The param setter becomes redundant using this approach.

There were multiple problems when trying to add another algorithm with a different name, some of these include
1) Using a different name means that the existing code then has to translate from HMAC to HMAC-WITH-SALT in multiple places
2) The property query probably needed to be fips=no, which then means you also have to change the same places in 1) to modify the query.
3) EVP_MAC_is_a(md, "HMAC") wouldnt have worked nicely without special casing it also.

This PR does solve these particular problems..

File: providers/fips/fipsprov.c
Comment (paulidale) [inline providers/fips/fipsprov.c:459]: Don't indent the `#` characters.

File: providers/fips/fipsprov.c
Comment (paulidale) [inline providers/fips/fipsprov.c:653]: Is it worth trimming this down to just the ones that are required?
Digest, cipher, MAC and rands at a guess.

**Review by paulidale — COMMENTED** [review]
Looks like a very credible solution to the problem.

File: providers/fips/fipsprov.c
Comment (slontis) [reply providers/fips/fipsprov.c:653]: Dont think it really matters..

Comment (paulidale) [reply providers/fips/fipsprov.c:653]: Not a lot I guess.  I was thinking that it would highlight which algorithms are internally fetched (& thus could have an _internal_ flag meaningfully added).

Comment (slontis) [reply providers/fips/fipsprov.c:653]: We have to be careful with this. It is useful when the higher level needs to do something different to the base,
not sure doing it for digest, ciphers is a good idea if you have to reproduce the checks in the higher level everywhere.

File: test/recipes/30-test_evp_data/evpmac_common.txt
Comment (paulidale) [inline test/recipes/30-test_evp_data/evpmac_common.txt:325]: There should be a test verifying the "unapproved" functionality too.

File: util/mk-fipsmodule-cnf.pl
Comment (paulidale) [inline util/mk-fipsmodule-cnf.pl:73]: not sure this is great placement amongst the KDFs.

File: apps/fipsinstall.c
Comment (paulidale) [inline apps/fipsinstall.c:109]: Move out of the "kdf" block perhaps?

File: util/mk-fipsmodule-cnf.pl
Comment (slontis) [reply util/mk-fipsmodule-cnf.pl:73]: trying to avoid conflicts for now.. will reorder all of them later

File: apps/fipsinstall.c
Comment (slontis) [reply apps/fipsinstall.c:109]: same reason

File: test/recipes/30-test_evp_data/evpmac_common.txt
Comment (slontis) [reply test/recipes/30-test_evp_data/evpmac_common.txt:325]: yep - need to test quite a few things including sskdf

**Comment by t8m** [comment]
> 4\. Add the concept of algorithm _usage_ via a settable OSSL_PARAM.  This would be user facing (& thus open to abuse) but it would address a number of other concerns.  E.g. MAC being used for verification has different FIPS restrictions than one being used for signing.

I like this solution the most.

File: test/recipes/30-test_evp_data/evpmac_common.txt
Comment (slontis) [reply test/recipes/30-test_evp_data/evpmac_common.txt:325]: This particular test fails in default also, so this is not related to a fips indicator.

**Comment by slontis** [comment]
> > 4. Add the concept of algorithm _usage_ via a settable OSSL_PARAM.  This would be user facing (& thus open to abuse) but it would address a number of other concerns.  E.g. MAC being used for verification has different FIPS restrictions than one being used for signing.
> 
> I like this solution the most.

The 'usage' will be a separate PR.
The internal query used here is still useful as a mechanism to not do the check in a child by default.

**Comment by slontis** [comment]
Fixed code style except for what it wants to do to the macro lines..

**Comment by slontis** [comment]
Looks like having a internal query that is different is causing some issues..

File: providers/implementations/macs/hmac_prov.c
Comment (paulidale) [inline providers/implementations/macs/hmac_prov.c:69]: Captial _I_ at the start :grinning:

File: doc/man7/provider-mac.pod
Comment (paulidale) [inline doc/man7/provider-mac.pod:200]: Line breaks are the best way to determine the single or double space between sentences.
I.e. if you're going to change it, make it a new line for both here.

**Review by paulidale — CHANGES_REQUESTED** [review]

**Comment by paulidale** [comment]
And an inevitable conflict which prevents the unit tests running.

File: providers/implementations/macs/hmac_prov.c
Comment (slontis) [reply providers/implementations/macs/hmac_prov.c:69]: Used 'internal' instead - I want it to match the variable.

File: apps/fipsinstall.c
Comment (paulidale) [inline apps/fipsinstall.c:642]: Spurious semicolon.

File: doc/man1/openssl-fipsinstall.pod.in
Comment (paulidale) [inline doc/man1/openssl-fipsinstall.pod.in:228]: The wording for these two options ought to be the same apart from the `H` and `K`.

File: doc/man7/provider-mac.pod
Comment (paulidale) [inline doc/man7/provider-mac.pod:199]: Line breaks are the best way to determine the single or double space between sentences.
I.e. if you're going to change it, make it a new line instead.

File: providers/implementations/macs/kmac_prov.c
Comment (paulidale) [inline providers/implementations/macs/kmac_prov.c:271]: Why the switch around here?

**Review by paulidale — COMMENTED** [review]
Looks good.

File: doc/man7/provider-mac.pod
Comment (slontis) [reply doc/man7/provider-mac.pod:199]: Removed the  space. Dont care about If it is a newline or not

File: providers/implementations/macs/kmac_prov.c
Comment (slontis) [reply providers/implementations/macs/kmac_prov.c:271]: It was removed then added back..

**Comment by slontis** [comment]
Rebased and NIT fixes..

**Review by paulidale — APPROVED** [review]

**Review by xnox — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
Needs a rebase.

**Comment by slontis** [comment]
@nhorman Can you get someone to review this PR please.

**Comment by t8m** [comment]
Needs rebase again

**Comment by slontis** [comment]
rebased again with no changes

**Comment by slontis** [comment]
ping @openssl/committers

File: include/openssl/core.h
Comment (beldmit) [inline include/openssl/core.h:45]: Do we need this change?

Comment (slontis) [reply include/openssl/core.h]: Removed - I was trying to fix code style nits.

**Comment by paulidale** [comment]
reping @openssl/committers

File: providers/fips/fipsprov.c
Comment (mattcaswell) [inline providers/fips/fipsprov.c:726]: Given that this function is identical to `fips_query` with the exception of this one line, would it be better have one call the other - with this function just handling `OSSL_OP_MAC` independently? I worry that future updates will cause these two functions to get out of sync with each other.

File: providers/implementations/macs/hmac_prov.c
Comment (Sashan) [inline providers/implementations/macs/hmac_prov.c:67]: I have a question:
is it OK to change structure size when code gets compiled with FIPS *disabled*?
I'm just worried about binary compatibility. The offset of members further below will also change.
but it does not matter perhaps.

also hmac_dup() seems to forget to copy `internal` member from src to dst.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Overall this looks ok. One comment below.

**Review by Sashan — CHANGES_REQUESTED** [review]

File: providers/implementations/macs/hmac_prov.c
Comment (mattcaswell) [reply providers/implementations/macs/hmac_prov.c:67]: > is it OK to change structure size when code gets compiled with FIPS disabled?

This makes no difference. This code is compiled *twice* when configured with "enable-fips". It is compiled once with `FIPS_MODULE` not defined and the resulting code is incorporated into the default provider. It is compiled a second time with `FIPS_MODULE` defined and that code is incorporated into the fips provider. This second version of the code is entirely contained within the fips provider. There are no binary compatibility issues with this.

Comment (slontis) [reply providers/implementations/macs/hmac_prov.c:67]: Reasonable question though..

Comment (mattcaswell) [reply providers/implementations/macs/hmac_prov.c:67]: > also hmac_dup() seems to forget to copy internal member from src to dst.

This should be covered by this line:

```` c
    *dst = *src;
````

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
@paulidale - please reconfirm

File: providers/implementations/macs/hmac_prov.c
Comment (Sashan) [reply providers/implementations/macs/hmac_prov.c:67]: OK I see it now. but still would prefer a consistency between kmac and hmac (if possible).
but this is just my preference which I don't insist on.

**Review by Sashan — APPROVED** [review]
looks good to me.

**Review by paulidale — APPROVED** [review]

File: providers/implementations/macs/hmac_prov.c
Comment (paulidale) [reply providers/implementations/macs/hmac_prov.c:67]: Having the structures change between the two compilations isn't a binary compatibility issue as @mattcaswell noted.  It causes annoyance debugging but provides, what I think is, the significant benefit of ensuring that the FIPS provider/libcrypto boundary remains pristine.  If a structure accidentally crosses this boundary, things start failing.

The annoyance is worst for the random bit generation where both sides of the boundary need to be poked with simultaneously to figure out what's going on.  For other algorithms, it's just one side which is generally okay.

**Review by paulidale — APPROVED** [review]
Yep, still good.

File: providers/fips/fipsprov.c
Comment (paulidale) [reply providers/fips/fipsprov.c:719]: The relatively small number of algorithms that are internally fetched by the FIPS provider would have allowed a simplified of this but reusing the external code is a reasonable solution too.

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**
